File: LeetCode/Peng/279PerfectSquares.cpp
//Grow ideal, add square numbers
class Solution {
public:
    int numSquares(int n) {
        vector<int> records(n + 1, INT_MAX);
        for (int i = 1; i * i <= n; i++) {
            records[i * i] = 1;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; i + j * j <= n; j++) {
                records[i + j * j] = min(records[i + j * j], records[i] + 1);
            }
        }
        return records[n];
    }
};

File: LeetCode/Hao/279PerfectSquares.java
public class Solution {
    public int numSquares(int n) {
        int[] d = new int[n+1];
        d[1] = 1;
        
        for (int i = 1; i * i <= n; i ++) {
            d[i * i] = 1;
        }
        
        for (int i = 2; i <= n; i ++) {
            for (int j = 1; j * j <= i; j ++) {
                int ans = d[i - j * j] + 1;
                d[i] = (d[i] == 0) ? ans : Math.min(ans, d[i]);
            }
        }
        
        return d[n];
    }
}
